### PR TITLE
[Console] Add command resolver (proof of concept)

### PR DESCRIPTION
**Pull request by funivan**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #12063 |
| License | MIT |
| Doc PR | NA |

Hello. This is simple proof of concept for command resolver.
# Why?

On huge projects we should always initialize all commands, even if they are not used.
Some users want lazy-loading of commands. And there are a lot of ways how to do that. Store cache in files, db, initialize commands by unique id etc..
# How?

This PR introduce new interface `CommandResolverInterface`. Any user can implement logic and inject it to the application. With this interface developer can create commands _by request_ 
# BC ?

This interface is softly injected to this application. All tests pass. By default `CommandResolver` is simple holder for commands. 
# Current state

There are one problem: when we add `Command` to application, application add this command only if it is available. This logic is not covered in `CommandResolverInterface`. We put all responsibility to developer which Commands are available to application.

If it will be interesting to the team I can polish it (fix code style, possible change names of interface, write documentation etc.)

### CODE REVIEW

File: src/Symfony/Component/Console/Application.php [hunk 1]
```diff
@@ -87,6 +93,8 @@ public function __construct($name = 'UNKNOWN', $version = 'UNKNOWN')
         $this->helperSet = $this->getDefaultHelperSet();
         $this->definition = $this->getDefaultInputDefinition();
 
+        $this->commands = !empty($commandResolver) ? $commandResolver : new CommandResolver();
```
Comment (stof) [inline src/Symfony/Component/Console/Application.php]: use a ``null`` comparison instead

File: src/Symfony/Component/Console/Application.php [hunk 2]
```diff
@@ -60,7 +62,10 @@
  */
 class Application
 {
-    private $commands = array();
+    /**
+     * @var CommandResolverInterface
+     */
+    private $commands;
```
Comment (stof) [inline src/Symfony/Component/Console/Application.php]: ``commandResolver`` would be a better name

File: src/Symfony/Component/Console/Application.php [hunk 3]
```diff
@@ -386,12 +390,12 @@ public function add(Command $command)
      */
     public function get($name)
     {
-        if (!isset($this->commands[$name])) {
+        $command = $this->commands->get($name);
+
+        if (!isset($command)) {
```
Comment (stof) [inline src/Symfony/Component/Console/Application.php]: don't use ``isset`` to check for null

File: src/Symfony/Component/Console/CommandsResolver/CommandResolver.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+ * @author Ivan Shcherbak <dev@funivan.com>
+ */
+class CommandResolver implements CommandResolverInterface
+{
+    private $commands = array();
+
+  /**
+   * {@inheritdoc}
+   */
+  public function add(Command $command)
```
Comment (stof) [inline src/Symfony/Component/Console/CommandsResolver/CommandResolver.php]: wrong indentation

File: src/Symfony/Component/Console/CommandsResolver/CommandResolver.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  {
+      $this->commands[$command->getName()] = $command;
+
+      foreach ($command->getAliases() as $alias) {
+          $this->commands[$alias] = $command;
+      }
+
+      return $this;
```
Comment (stof) [inline src/Symfony/Component/Console/CommandsResolver/CommandResolver.php]: -1 for making it fluent

File: src/Symfony/Component/Console/CommandsResolver/CommandResolverInterface.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+{
+  /**
+   * @param Command $command
+   *
+   * @return $this
+   */
+  public function add(Command $command);
+
+  /**
+   * Check if command exist.
```
Comment (stof) [inline src/Symfony/Component/Console/CommandsResolver/CommandResolverInterface.php]: ``checks whether the command exists``

File: src/Symfony/Component/Console/CommandsResolver/CommandResolver.php (hunk 5 above)
Comment (funivan) [inline src/Symfony/Component/Console/CommandsResolver/CommandResolver.php]: @stof can you explain why? 
Should I return Command object instead ?

Comment (stof) [inline src/Symfony/Component/Console/CommandsResolver/CommandResolver.php]: IMO, you should not return anything. Returning the command object does not make sense: if the user need it, they can save it first as they are passing it as argument.

Btw, the code using this method ignores the return value entirely, which is a clear indication that there is no meaningful one.

and regarding the fluent interface, I suggest you to read http://ocramius.github.io/blog/fluent-interfaces-are-evil/ as there are good arguments in it

Comment (funivan) [inline src/Symfony/Component/Console/CommandsResolver/CommandResolver.php]: Ok thanks. I will fix it in a minute.

File: src/Symfony/Component/Console/CommandsResolver/CommandResolverInterface.php (hunk 6 above)
Comment (funivan) [inline src/Symfony/Component/Console/CommandsResolver/CommandResolverInterface.php]: Add method should check if command exist?

File: src/Symfony/Component/Console/Application.php [hunk 7]
```diff
@@ -87,6 +93,8 @@ public function __construct($name = 'UNKNOWN', $version = 'UNKNOWN')
         $this->helperSet = $this->getDefaultHelperSet();
         $this->definition = $this->getDefaultInputDefinition();
 
+        $this->commandResolver = ($commandResolver !== null) ? $commandResolver : new CommandResolver();
```
Comment (aitboudad) [inline src/Symfony/Component/Console/Application.php]: it should be ```null !== $commandResolver  ?``` and parentheses is useless here

File: src/Symfony/Component/Console/CommandsResolver/CommandResolver.php [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function has($name)
+    {
+        return !empty($this->commands[$name]);
```
Comment (aitboudad) [inline src/Symfony/Component/Console/CommandsResolver/CommandResolver.php]: `isset` is enough

File: src/Symfony/Component/Console/CommandsResolver/CommandResolver.php [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function get($name)
+    {
+        return !empty($this->commands[$name]) ? $this->commands[$name] : null;
```
Comment (aitboudad) [inline src/Symfony/Component/Console/CommandsResolver/CommandResolver.php]: same here

File: src/Symfony/Component/Console/CommandsResolver/CommandResolver.php (hunk 8 above)
Comment (funivan) [inline src/Symfony/Component/Console/CommandsResolver/CommandResolver.php]: fixed

File: src/Symfony/Component/Console/Tests/CommandResolver/CommandResolverTest.php [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+use Symfony\Component\Console\Application;
+use Symfony\Component\Console\Tests\Fixtures\CustomCommandResolver;
+use Symfony\Component\Console\Tests\Fixtures\LazyTestCommand;
+
+/**
+ * @author Ivan Shcherbak <dev@funivan.com>
+ */
+class CommandResolverTest extends \PHPUnit_Framework_TestCase
+{
+  public function testLazyCommandResolver()
```
Comment (stloyd) [inline src/Symfony/Component/Console/Tests/CommandResolver/CommandResolverTest.php]: Wrong indent in whole file. Use 4 spaces not 2.

File: src/Symfony/Component/Console/Tests/Fixtures/CustomCommandResolver.php [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+namespace Symfony\Component\Console\Tests\Fixtures;
+
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\CommandsResolver\CommandResolverInterface;
+
+/**
+ * @author Ivan Shcherbak <dev@funivan.com>
+ */
+class CustomCommandResolver implements CommandResolverInterface
```
Comment (stloyd) [inline src/Symfony/Component/Console/Tests/Fixtures/CustomCommandResolver.php]: Why not extend existing one and just add default fixture code in constructor?

Comment (funivan) [inline src/Symfony/Component/Console/Tests/Fixtures/CustomCommandResolver.php]: I want to show how to implement `CommandResolverInterface` from zero.
What do you mean: `add default fixture code in constructor`? Can you show me an example, I don`t understand.

File: src/Symfony/Component/Console/Tests/CommandResolver/CommandResolverTest.php (hunk 10 above)
Comment (funivan) [inline src/Symfony/Component/Console/Tests/CommandResolver/CommandResolverTest.php]: Sorry for that. I will fix it.

**Comment by funivan** [comment]
@stof Hello. Can you help me? I can`t pass all tests of this PR at CI services (In local environment all tests are green). 
Some times it failure in AppVeyor sometimes in travis, but always in different places. And even more, sometimes failures are not relevant to my commits. Thank you. 
What should I do to make this PR green?

**Comment by TomasVotruba** [comment]
Very nice PR!

Tests for Console are passing well. It fails on Symfony/Bridge/ProxyManager package, not sure why.

**Comment by funivan** [comment]
@TomasVotruba yep and I don`t know how to resolve this. Is there any way to rerun this tests?

**Comment by DavidBadura** [comment]
with a rebase :wink:

**Comment by funivan** [comment]
Hi 2 all. It is easier to create new pull request and implement the same logic. 
If this feature is interesting to the community and to the core time I can implement the same algorithm of the command resolver.

File: src/Symfony/Component/Console/Application.php [hunk 12]
```diff
@@ -76,17 +81,20 @@ class Application
...
-    public function __construct($name = 'UNKNOWN', $version = 'UNKNOWN')
+    public function __construct($name = 'UNKNOWN', $version = 'UNKNOWN', CommandResolverInterface $commandResolver = null)
     {
         $this->name = $name;
         $this->version = $version;
         $this->defaultCommand = 'list';
         $this->helperSet = $this->getDefaultHelperSet();
         $this->definition = $this->getDefaultInputDefinition();
 
+        $this->commandResolver = $commandResolver !== null ? $commandResolver : new CommandResolver();
```
Comment (chalasr) [inline src/Symfony/Component/Console/Application.php]: I think we use `$this->commandResolver = $commandResolver ?: new CommandResolver()`

**Comment by nicolas-grekas** [comment]
Closing in favor of #22734 which is going to be accepted and provide most of what this tries to achieve.
Thanks for working on this @funivan.

**Pull request closed without merging**